go/vt/vtctl/grpcvtctldserver/server.go

```diff
@@ -2033,6 +2035,228 @@ func (s *VtctldServer) UpdateCellsAlias(ctx context.Context, req *vtctldatapb.Up
+	shards, err := s.ts.GetShardNames(ctx, req.Keyspace)
+	if err != nil {
+		return nil, vterrors.Wrapf(err, "GetShardNames(%v) failed: %v", req.Keyspace, err)
+	}
+
+	sort.Strings(shards)
+	if len(shards) == 1 {
+		validateSchemaShardReq := &vtctldatapb.ValidateSchemaShardRequest{
+			Keyspace:       req.Keyspace,
+			ExcludeTables:  req.ExcludeTables,
+			Shard:          shards[0],
+			IncludeVSchema: req.IncludeVSchema,
+			IncludeViews:   req.IncludeViews,
+		}
+
+		if _, err := s.ValidateSchemaShard(ctx, validateSchemaShardReq); err != nil {
+			return nil, vterrors.Wrapf(err, "ValidateSchemaShard(%v) failed: %v", validateSchemaShardReq, err)
+		}
+		return &vtctldatapb.ValidateSchemaKeyspaceResponse{}, nil
+	}
+
+	var (
+		referenceSchema *tabletmanagerdatapb.SchemaDefinition
+		referenceAlias  *topodatapb.TabletAlias
+		rec             concurrency.AllErrorRecorder
+		wg              sync.WaitGroup
+	)
+
+	if req.IncludeVSchema {
+		validateVSchemaReq := &vtctldatapb.ValidateVSchemaRequest{
+			Keyspace:      req.Keyspace,
+			Shards:        shards,
+			IncludeViews:  req.IncludeViews,
+			ExcludeTables: req.ExcludeTables,
+		}
+
+		if _, err := s.ValidateVSchema(ctx, validateVSchemaReq); err != nil {
+			return nil, vterrors.Wrapf(err, "ValidateVSchema(%v) failed: %v", validateVSchemaReq, err)
+		}
+	}
+
+	// then diffs all tablets in the other shards
+	for _, shard := range shards {
+		si, err := s.ts.GetShard(ctx, req.Keyspace, shard)
+		if err != nil {
+			rec.RecordError(fmt.Errorf("GetShard(%v, %v) failed: %v", req.Keyspace, shard, err))
+			continue
+		}
+
+		if !si.HasPrimary() {
+			if !req.SkipNoPrimary {
+				rec.RecordError(fmt.Errorf("no primary in shard %v/%v", req.Keyspace, shard))
+			}
+			continue
+		}
+
+		if referenceSchema == nil {
+			referenceAlias = si.Shard.PrimaryAlias
+			log.Infof("Gathering schema for reference primary %v", topoproto.TabletAliasString(referenceAlias))
+
+			getTabletReq := &vtctldata.GetTabletRequest{
+				TabletAlias: referenceAlias,
+			}
+
+			getTabletResp, err := s.GetTablet(ctx, getTabletReq)
+			if err != nil {
+				return nil, vterrors.Wrapf(err, "GetTablet(%v) failed: %v", referenceAlias, err)
+			}
+
+			referenceSchema, err = s.tmc.GetSchema(ctx, getTabletResp.GetTablet(), nil, req.ExcludeTables, req.IncludeViews)
+			if err != nil {
+				return nil, vterrors.Wrapf(err, "GetSchema(%v, nil, %v, %v) failed: %v", referenceAlias, req.ExcludeTables, req.IncludeViews, err)
+			}
+		}
+
+		aliases, err := s.ts.FindAllTabletAliasesInShard(ctx, req.Keyspace, shard)
+		if err != nil {
+			rec.RecordError(fmt.Errorf("FindAllTabletAliasesInShard(%v, %v) failed: %v", req.Keyspace, shard, err))
+			continue
+		}
+
+		for _, alias := range aliases {
+			// Don't diff schemas for self
+			if referenceAlias == alias {
+				continue
+			}
+			wg.Add(1)
+			go s.diffSchema(ctx, referenceSchema, referenceAlias, alias, req.ExcludeTables, req.IncludeViews, &wg, &rec)
```

Comment:
rather than passing through the waitgroup and recorder, let's keep the concurrency logic focused here rather than in two places:

```go
wg.Add(1)
go func(/*params as needed*/) {
    defer wg.Done()
    if err := s.diffSchema(/*params*/); err != nil {
        rec.RecordError(err)
    }
}(/*params*/)
```